### PR TITLE
Enable npm as package manager

### PR DESCRIPTION
**Pull request by Fatme**
Should be merged after this one https://github.com/telerik/mobile-cli-lib/pull/351

### CODE REVIEW

File: lib/definitions/semver.d.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+
+declare module "semver" {
```
Comment (rosen-vladimirov) [inline lib/definitions/semver.d.ts]: I think we already have the semver definitions in common

File: lib/services/platform-service.ts [hunk 2]
```diff
@@ -137,7 +140,7 @@ export class PlatformService implements IPlatformService {
 
 	public preparePlatform(platform: string): IFuture<void> {
 		return (() => {
-			platform = platform.toLowerCase();
+		 	platform = platform.toLowerCase();
```
Comment (teobugslayer) [inline lib/services/platform-service.ts]: Is this added space needed?

File: lib/services/platform-service.ts [hunk 3]
```diff
@@ -168,14 +171,16 @@ export class PlatformService implements IPlatformService {
...
 
+			var tnsModulesDestinationPath = path.join(platformData.appDestinationDirectoryPath, constants.APP_FOLDER_NAME, PlatformService.TNS_MODULES_FOLDER_NAME);
+			this.$broccoliBuilder.prepareNodeModules(tnsModulesDestinationPath, this.$projectData.projectDir).wait();
+
 			this.$logger.out("Project successfully prepared");
 
 		}).future<void>()();
 	}
-
+
```
Comment (teobugslayer) [inline lib/services/platform-service.ts]: Is this change needed?

File: lib/tools/broccoli/broccoli-plugin-wrapper.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  constructor(private pluginClass: any, wrappedPluginArguments: any) { 
+      this.inputTree = wrappedPluginArguments[0];
+      this.description = this.pluginClass.name;
+      this.absoluteOutputPath = wrappedPluginArguments[1];
+      this.treeRootDirName = wrappedPluginArguments[2];
+      this.projectDir = wrappedPluginArguments[3];
+      this.$injector = $injector.resolve("injector");
+  }
+
+  rebuild(): any {
```
Comment (teobugslayer) [inline lib/tools/broccoli/broccoli-plugin-wrapper.ts]: The code does not return but function is declared to return any. Consider declaring it as void.

File: lib/tools/broccoli/builder.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+  private rebuild(builder: any): IFuture<any> {
+    let future = new Future<any>();
+    builder.build()
+        .then((result: any) => {
+              future.return(result);
+        })
+        .catch((err: any) => {
+              if(err.file) {
+                console.error("File: " + err.file);
```
Comment (teobugslayer) [inline lib/tools/broccoli/builder.ts]: Looks like debug code to me. Is it needed? If yes, consider using $loggger instead.

File: lib/tools/broccoli/node-modules-dest-copy.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		  let dirs = fs.readdirSync(this.outputRoot);
+		  _.each(dirs, dir => {
+			  let fileContent = require(path.join(dir, constants.PACKAGE_JSON_FILE_NAME));
+			  result[fileContent.name] = fileContent;
+			 });
+	  } catch(err) {
+		  result = Object.create(null);
+	  }
+	  
+	  return result;
```
Comment (teobugslayer) [inline lib/tools/broccoli/node-modules-dest-copy.ts]: As the function is written, it silently skips all dependencies in case any of them is corrupted. This means that code after that will randomly fail. Maybe it is better to let the exception cancel the current operation so that the user can have contextual information what's wrong and can fix it.

File: lib/tools/broccoli/trees/node-modules-tree.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+///<reference path="../../../.d.ts"/>
+"use strict";
+
+let Funnel = require('broccoli-funnel');
+let path = require("path");
```
Comment (teobugslayer) [inline lib/tools/broccoli/trees/node-modules-tree.ts]: this can be import'ed

**Comment by teobugslayer** [comment]
You need to rebase on master

File: lib/definitions/semver.d.ts (hunk 1 above)
Comment (Fatme) [inline lib/definitions/semver.d.ts]: Yes, they were moved to common.

**Comment by ErjanGavalji** [comment]
:+1:

File: lib/tools/broccoli/builder.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    let future = new Future<any>();
+    builder.build()
+        .then((result: any) => {
+              future.return(result);
+        })
+        .catch((err: any) => {
+              if(err.file) {
+                this.$logger.error("File: " + err.file); 
+              }
+              if(err.statck) {
```
Comment (teobugslayer) [inline lib/tools/broccoli/builder.ts]: maybe err.stack? (a spelling mistake)

File: lib/tools/broccoli/builder.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+                this.$logger.error(err.stack);
+              }
+              future.throw(err.toString());
```
Comment (teobugslayer) [inline lib/tools/broccoli/builder.ts]: Why are you throwing a string instead of an Error?

File: lib/tools/broccoli/node-modules-dest-copy.ts [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+				};
+				
+				let addedDependency = dependencies[currentDependency.name];
+				if (addedDependency) {
+					if (semver.gt(currentDependency.version, addedDependency.version)) {
+						let currentDependencyMajorVersion = semver.major(currentDependency.version);
+						let addedDependencyMajorVersion = semver.major(addedDependency.version);
+		
+						let message = util.format("Old version %s, new version %s. Old location %s, new location %s.", addedDependency.version, currentDependency.version, addedDependency.directory, currentDependency.directory);
+						currentDependencyMajorVersion === addedDependencyMajorVersion ? console.log(message) : console.log(message);
```
Comment (teobugslayer) [inline lib/tools/broccoli/node-modules-dest-copy.ts]: This code does not look write - the if and else code is the same.

File: lib/services/plugins-service.ts [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+				}
+			};
+			
+			this.executeForAllInstalledPlatforms(action);
+		}).future<void>()();
+	}
+	
+	public getAllInstalledPlugins(): IFuture<IPluginData[]> {
+		return (() => {
+			let nodeModules = this.$fs.readDirectory(this.nodeModulesPath).wait();
```
Comment (rosen-vladimirov) [inline lib/services/plugins-service.ts]: make sure `this.nodeModulesPath` directory exists

**Comment by rosen-vladimirov** [comment]
In case you remove node_modules dir manually, the plugin is not added to the project after build. I suggest to call `npm install` in project dir on prepare step. This way all dependencies will be installed in node_modules dir before building the project. 
Add/Remove plugin commands can check the package.json's dependencies instead of listing node_modules

File: lib/services/plugins-service.ts [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+	}
+	
+	private parseNpmCommandResult(npmCommandResult: string): string { // The npmCommandResult is in the following format: [<name>@<version node_modules/<name>]
+		return npmCommandResult.split("@")[0]; // returns plugin name
+	}
+	
+	private executeForAllInstalledPlatforms(action: (pluginDestinationPath: string, platform: string, platformData: IPlatformData) => void): void {
+		let availablePlatforms = _.keys(this.$platformsData.availablePlatforms);
+		_.each(availablePlatforms, platform => {
+			let isPlatformInstalled = this.$fs.exists(path.join(this.$projectData.platformsDir, platform.toLowerCase())).wait();
```
Comment (teobugslayer) [inline lib/services/plugins-service.ts]: There is such function in platform service but it is private. Maybe we can reuse it here?

File: lib/services/platform-service.ts [hunk 13]
```diff
@@ -305,6 +315,10 @@ export class PlatformService implements IPlatformService {
 			}
 		}).future<void>()();
 	}
+	
+	private isPlatformInstalled(platform: string): IFuture<boolean> {
```
Comment (teobugslayer) [inline lib/services/platform-service.ts]: It seems that platform name is not always lower-cased. Consider adding toLowerCase().

File: lib/services/plugins-service.ts (hunk 12 above)
Comment (Fatme) [inline lib/services/plugins-service.ts]: Actually I wanted to reuse it. The problem is that `pluginsService` is dependency of `platformsService` so here I am not able to inject `pluginsService` :smile:

File: lib/services/platform-service.ts (hunk 13 above)
Comment (Fatme) [inline lib/services/platform-service.ts]: Done :)

File: lib/declarations.ts [hunk 14]
```diff
@@ -39,7 +39,8 @@ interface IOptions extends ICommonOptions {
 	emulator: boolean;
 	symlink: boolean;
 	forDevice: boolean;
-	client: boolean;
+	client: boolean
```
Comment (rosen-vladimirov) [inline lib/declarations.ts]: missing `;` after client

File: lib/services/platform-service.ts [hunk 15]
```diff
@@ -169,14 +177,17 @@ export class PlatformService implements IPlatformService {
...
+			// Process node_modules folder
+			this.$pluginsService.ensureAllDependenciesAreInstalled().wait();
+			var tnsModulesDestinationPath = path.join(platformData.appDestinationDirectoryPath, constants.APP_FOLDER_NAME, PlatformService.TNS_MODULES_FOLDER_NAME);
+			this.$broccoliBuilder.prepareNodeModules(tnsModulesDestinationPath, this.$projectData.projectDir).wait();
 
+			this.$logger.out("Project successfully prepared");
 		}).future<void>()();
 	}
-
+
```
Comment (rosen-vladimirov) [inline lib/services/platform-service.ts]: some unneeded tabs :)

File: lib/services/plugins-service.ts (hunk 12 above)
Comment (teobugslayer) [inline lib/services/plugins-service.ts]: We discussed in person and agreed to leave it as-is.
